compiler/aarch64/codegen/OMRMemoryReference.cpp

```diff
@@ -444,18 +549,82 @@ void OMR::ARM64::MemoryReference::populateMemoryReference(TR::Node *subTree, TR:
          TR::Node *addressChild = subTree->getFirstChild();
          TR::Node *integerChild = subTree->getSecondChild();
 
-         self()->populateMemoryReference(addressChild, cg);
-         intptr_t amount = (integerChild->getOpCodeValue() == TR::iconst) ?
-                           integerChild->getInt() : integerChild->getLongInt();
-         self()->addToOffset(integerChild, amount, cg);
-         cg->decReferenceCount(integerChild);
+         if (subTree->getSecondChild()->getOpCode().isLoadConst())
+            {
+            // array access with constant index
+            self()->populateMemoryReference(addressChild, cg);
+            intptr_t amount = (integerChild->getOpCodeValue() == TR::iconst) ?
+                              integerChild->getInt() : integerChild->getLongInt();
+            self()->addToOffset(integerChild, amount, cg);
+            if (comp->getOption(TR_TraceCG))
+               {
+               traceMsg(comp, "Capturing array access with constant index at node %p offset = %d\n", subTree, amount);
+               }
+            cg->decReferenceCount(integerChild);
+            }
+         else if (cg->whichNodeToEvaluate(addressChild, integerChild) == 1)
+            {
+            self()->populateMemoryReference(integerChild, cg);
+
+            self()->populateMemoryReference(addressChild, cg);
+            }
+         else
+            {
+            self()->populateMemoryReference(addressChild, cg);
+
+            if (_baseRegister != NULL && _indexRegister != NULL)
+               {
+               self()->consolidateRegisters(subTree, cg);
+               }
+
+            self()->populateMemoryReference(integerChild, cg);
+            }
+         cg->decReferenceCount(subTree);
+         }
+      else if ((subTree->getOpCodeValue() == TR::lsub) &&
+               (subTree->getSecondChild()->getOpCodeValue() == TR::lconst) &&
+               checkOffset(subTree->getSecondChild(), cg, -subTree->getSecondChild()->getLongInt(), _length))
+         {
+         TR::Node *constChild = subTree->getSecondChild();
+         intptr_t amount = -constChild->getLongInt();
+         self()->populateMemoryReference(subTree->getFirstChild(), cg);
+
+         self()->addToOffset(subTree, amount, cg);
+         if (comp->getOption(TR_TraceCG))
+            {
+            traceMsg(comp, "Capturing lsub node with constant value at node %p offset = %d\n", subTree, amount);
+            }
+         cg->decReferenceCount(constChild);
+         cg->decReferenceCount(subTree);
+         }
+      else if (subTree->getOpCodeValue() == TR::i2l)
+         {
+         if (_indexRegister != NULL)
+            {
+            if ((_baseRegister != NULL) || (_scale != 0) || self()->isIndexSignExtended())
+               {
+               self()->consolidateRegisters(subTree, cg);
+               }
+            else
+               {
+               _baseRegister = _indexRegister;
+               _baseNode     = _indexNode;
```

Comment:
Is there any reason for not using `moveIndexToBase()` here?

Reply:
Good catch.👍
I changed `moveIndexToBase` to cover the case when `_baseRegister` is not NULL and simplified the code to free `_indexRegister` in `populateMemoryReference` function.